Add sandbox dockerfile generator script

marwan37 opened the pull request:
This PR adds:

1. A python script for generating Dockerfiles for use within the upcoming zenml sandbox environment.
2. A GH Actions workflow that automatically builds and pushes Docker images to `zenmldocker` on DockerHub when changes are made to project directories.

**Changes:**
- Added python script to generate standardized Dockerfile.codespace files for our projects
- Moved our python scripts to the `scripts` directory
- Used to script to generate a `Dockerfile.codespace` for oncoclear and omni-reader as examples

**How the workflow works:**
- Detects and builds Docker images only for projects with changes
- Uses `path-ignore` pattern to exclude non-project files and directories
- For projects without a `Dockerfile.codespace`:
  - Generates one using the `generate_codespace_dockerfile.py` script
  - Creates a PR for review instead of committing directly
  - Only builds and pushes images after the PR is merged
- For projects with an existing `Dockerfile.codespace`:
  - Automatically builds and pushes to the latest tag on DockerHub
- Images are automatically tagged with a **timestamp** to preserve older project images (if any)

strickvl commented on oncoclear/Dockerfile.sandbox:
same question about `uv` here

strickvl commented on scripts/generate_sandbox_dockerfile.py:
```diff
@@ -0,0 +1,224 @@
+LABEL project_name="{project_name}"
+LABEL project_version="0.1.0"
+
+# Install project-specific dependencies
+RUN pip install --no-cache-dir \\
+{dependencies}
```
here I'd do the uv stuff first. (passing `--system` etc or whatever is needed to setup the environment inside these sandboxes.

marwan37 replied:
for sure.

strickvl commented on scripts/generate_sandbox_dockerfile.py:
```diff
@@ -0,0 +1,224 @@
+        for line in f:
+            line = line.strip()
+            if line and not line.startswith("#"):
+                if line.startswith("polars"):
+                    line = line.replace("polars", "polars-lts-cpu")
```
this probably needs a comment to explain why the exception...

strickvl commented on scripts/generate_sandbox_dockerfile.py:
```diff
@@ -0,0 +1,224 @@
+    return dependencies
+
+
+def detect_api_keys(project_dir):
+    """Attempt to detect required API keys by scanning Python files."""
+    api_patterns = {
```
This feels like maybe a bit of an antipattern! Or if not, then there needs to be some sort of checklist where it's explained that if you add a new ENV variable in your project then probably you need to update this script etc?

marwan37 replied:
good catch, will update this to just read the `.env` file.

marwan37 commented on scripts/generate_sandbox_dockerfile.py:
```diff
@@ -0,0 +1,224 @@
+
+def parse_requirements(project_dir):
+    """Parse requirements.txt file if it exists."""
+    req_file = Path(project_dir) / "requirements.txt"
```
thought we talked about normalizing the projects to all use `requirements.txt`? In any case, it shouldn't be too complicated of an updated, but we'll see.

marwan37 commented on omni-reader/Dockerfile.sandbox:
yes!  👍 initial script added `uv` to the requirements, but removed it as i wasn't sure we were aligned.

marwan37 commented:
@strickvl I've updated the script to use `uv` for installing deps. For env variables, the script now creates a `.env` file in the Dockerfile only if it exists in the project - otherwise users will need to set API keys via env variables as described in each project's README. 

I've also renamed the base docker image to `zenmldocker/zenml-projects:base` and was thinking we can use a consistent naming convention for projects: `zenmldocker/zenml-projects:oncoclear`, `zenmldocker/zenml-projects:omni-reader`, etc. Let me know your thoughts.

strickvl commented:
That naming convention feels a bit weird. So you're using the tag as the name instead of the actual image name? How would we support multiple versions of a project? Just when we update / upgrade we no longer offer a supported version?

marwan37 commented:
@strickvl hmm I was thinking we'd have 1 version at most for each project, but yes the naming convention does feel weird. will revert to `zenmldocker/zenml-sandbox` then, and `zenmldocker/projects-oncoclear`, etc. for projects.

strickvl commented:
Practically we'd probably only support one version. But I think in practice, some people might be playing around with one version etc and then if we 'upgrade' that version (as we have done with our projects many times) then suddenly that person's code / run behaviour changes etc.. I think that's probably undesirable.

marwan37 commented:
@strickvl Gotcha, though adding custom tags to the build/push docker images workflow might get tricky. When we need to preserve the original image while pushing a new tag, we'd either need to manually trigger the workflow with a `tag` input or use a versioning system (like starting with `base` and incrementing with each change). For now, I've kept it simple with just the "latest" tag -- and we can expand this later if needed?

htahir1 commented on omni-reader/Dockerfile.sandbox:
I wouldnt call this "sandbox" , id rather call it something like "Codespace"

htahir1 commented on scripts/generate_sandbox_dockerfile.py:
name to generate_codespace_dockerfile.py

htahir1 commented on scripts/generate_zenml_project.py:
we can delete this file it doesnt need to exist

htahir1 reviewed (changes requested):
I also added to the naming discussion above. overall looks good though

marwan37 commented:
I've renamed the base image to `zenmldocker/zenml-sandbox` and projects now follow this naming pattern:
- `zenmldocker/projects-oncoclear`
- `zenmldocker/projects-omni-reader`
- and so on...

strickvl commented on pyproject.toml:
```diff
@@ -6,6 +6,7 @@ authors = ["ZenML CodeMonkey <codemonkey@zenml.io>"]
 
 [tool.poetry.dependencies]
 python = ">=3.7.0,<3.9.0"
+tomli = "^2.0.1"  # Added for parsing pyproject.toml files when generating Dockerfile.codespace
```
https://pypi.org/project/tomli/ if you read the README, it seems to suggest that this is now part of the standard library now and that `tomli` is mainly for old versions of Pythno. So do we need it?

marwan37 replied:
It was added to python 3.11, but `zenml-projects` is set to `python = ">=3.7.0,<3.9.0"`. Unless we update this, we'd still need to include `tomli` to parse `pyproject.toml` files. Do you know if there was a specific reason for these python versions?

marwan37 replied:
zenml 0.81 supports `python = ">=3.9,<3.13"`, so I can adjust it to `">=3.11.0,<3.13.0"`.

strickvl replied:
for the whole of projects?

marwan37 replied:
just for the base `zenml-projects`, individual projects can still set their own python versions.
